io-engine/src/rebuild/nexus_rebuild.rs

```diff
@@ -0,0 +1,279 @@
+use snafu::ResultExt;
+use spdk_rs::LbaRange;
+use std::{
+    ops::{Deref, Range},
+    rc::Rc,
+};
+
+use crate::{
+    core::{DescriptorGuard, UntypedBdev},
+    gen_rebuild_instances,
+    rebuild::{
+        rebuild_error::{RangeLockFailed, RangeUnlockFailed},
+        rebuild_task::{RebuildTask, RebuildTaskCopier},
+    },
+};
+
+use super::{
+    rebuild_descriptor::RebuildDescriptor,
+    rebuild_error::{BdevNotFound, RebuildError},
+    rebuild_job::RebuildJob,
+    rebuild_job_backend::RebuildBackend,
+    rebuild_task::{RebuildTasks, TaskResult},
+    RebuildJobOptions,
+    SEGMENT_TASKS,
+};
+
+/// A Nexus rebuild job is responsible for managing a rebuild (copy) which reads
+/// from source_hdl and writes into destination_hdl from specified start to end.
+/// Each copy is synchronized with the nexus bdev using ranged locks to ensure
+/// that there is no concurrent between the nexus and the rebuild.
+/// This is a frontend interface that communicates with a backend runner which
+/// is the one responsible for the read/writing of the data.
+pub struct NexusRebuildJob(RebuildJob);
+
+impl std::fmt::Debug for NexusRebuildJob {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+impl Deref for NexusRebuildJob {
+    type Target = RebuildJob;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl NexusRebuildJob {
+    /// Creates a new RebuildJob which rebuilds from source URI to target URI
+    /// from start to end (of the data partition); notify_fn callback is called
+    /// when the rebuild state is updated - with the nexus and destination
+    /// URI as arguments.
+    /// todo: Should we use a builder? Example:
```

Comment:
@dsharma-dc @hrudaya21 @dsavitskiy 
What do you think, would this be good to have rather than a large new?

Reply:
As far I know, 5 parameters per function is industry standard. Mostly the future if there will be any enhancement here, We may extend more parameters to `RebuildOptions` which will keep the parameter list to 6.

My suggestion to keep 5 parameters rule, until We find some exceptions or complexity.